File: Semestrul 5/FLCD/lab-03/src/main/java/ubb/BinarySearchTree.java
package ubb;

public class BinarySearchTree {
    private static class Node {
        String value;
        int index; 
        Node left, right;

        Node(String value, int index) { 
            this.value = value;
            this.index = index;
        }
    }

    private Node root;
    private int nextIndex; 

    public BinarySearchTree() {
        this.root = null;
        this.nextIndex = 0; 
    }

    public void insert(String value) {
        root = insertRecursive(root, value);
    }

    private Node insertRecursive(Node node, String value) {
        if (node == null) {
            int newIndex = this.nextIndex;
            this.nextIndex++;
            return new Node(value, newIndex);
        }

        if (value.compareTo(node.value) < 0)
            node.left = insertRecursive(node.left, value);
        else if (value.compareTo(node.value) > 0)
            node.right = insertRecursive(node.right, value);
        return node;
    }

    public boolean contains(String value) {
        return containsRecursive(root, value);
    }

    private boolean containsRecursive(Node node, String value) {
        if (node == null) return false;
        if (value.equals(node.value)) return true;
        return value.compareTo(node.value) < 0
                ? containsRecursive(node.left, value)
                : containsRecursive(node.right, value);
    }


    public Integer get(String value) { 
        return getRecursive(root, value);
    }

    private Integer getRecursive(Node node, String value) { 
        if (node == null) {
            return null; 
        }

        if (value.equals(node.value)) {
            return node.index; 
        }

        return value.compareTo(node.value) < 0
                ? getRecursive(node.left, value)
                : getRecursive(node.right, value);
    }

    public void print() {
        printRecursive(root);
        System.out.println(); 
    }

    private void printRecursive(Node node) {
        if (node != null) {
            printRecursive(node.left);
            System.out.print(node.value + " " + node.index + "\n");
            printRecursive(node.right);
        }
    }
}

File: Semestrul 5/FLCD/lab-06/binary_search_tree.py
class BinarySearchTree:
    class _Node:
        def __init__(self, value, index):
            self.value = value
            self.index = index
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None
        self.next_index = 0

    def insert(self, value):
        self.root = self._insert_recursive(self.root, value)

    def _insert_recursive(self, node, value):
        if node is None:
            new_index = self.next_index
            self.next_index += 1
            return self._Node(value, new_index)

        if value < node.value:
            node.left = self._insert_recursive(node.left, value)
        elif value > node.value:
            node.right = self._insert_recursive(node.right, value)
        return node

    def contains(self, value):
        return self._contains_recursive(self.root, value)

    def _contains_recursive(self, node, value):
        if node is None:
            return False
        if value == node.value:
            return True

        return self._contains_recursive(node.left, value) if value < node.value \
            else self._contains_recursive(node.right, value)

    def get(self, value):
        return self._get_recursive(self.root, value)

    def _get_recursive(self, node, value):
        if node is None:
            return None
        if value == node.value:
            return node.index

        return self._get_recursive(node.left, value) if value < node.value \
            else self._get_recursive(node.right, value)

    def print_tree(self):
        self._print_recursive(self.root)
        print()

    def _print_recursive(self, node):
        if node is not None:
            self._print_recursive(node.right)
            print(f"{node.index}: {node.value} ")
            self._print_recursive(node.left)

